server/src/chatController.ts

```diff
@@ -0,0 +1,172 @@
+  next();
+};
+
+// GET / - list all chats for current user
+// Optional: ?vendor=<vendorId> filter
+ChatController.get('/', async (req, res) => {
+  const auth0Id = req.auth?.payload.sub;
+  const vendorFilter = req.query.vendor as string | undefined;
+
+  const query: any = {
+    $or: [
+      { 'user.id': auth0Id },
+      { 'vendor.id': auth0Id }
+    ]
+  };
+  if (vendorFilter) query['vendor.id'] = vendorFilter;
+
+  const chats = await Chat.find(query)
+    .sort({ updatedAt: -1 })
+    .populate('user', 'name')
+    .populate('vendor', 'name');
+
+  res.json(chats);
+});
+
+// GET /:id - get one chat by ID
+ChatController.get('/:id', verifyChatAccess, async (req, res) => {
+  //already verified chat access
+  //find the chat 
+  const inputId = req.params.id;
+  const chatObj = await Chat.findById({ inputId })
+  .populate(['user', 'vendor'])
+  .select('user.name vendor.name time messages invoices');
+  res.json(req.chat);
+});
+
+// POST /create - basic create chat
+ChatController.post('/create', async (req, res) => {
+  const { vendor } = req.body;
+
+  if (!vendor) {
+    res.status(400).json('Missing vendor');
+    return;
+  }
+
+  const auth0Id = req.auth?.payload.sub;
+  const userObj = await User.findOne({ auth0Id });
+
+  if (!userObj) {
+    res.status(404).json('Authenticated user not found');
+    return;
+  }
+
+  const vendorObj = await Vendor.findById(vendor.id);
+  if (!vendorObj) {
+    res.status(404).json('Vendor not found');
+    return; 
+  }
+
+  // Prevent duplicate chat between same user and vendor
+  const existingChat = await Chat.findOne({
+    'user.id': userObj.auth0Id,
+    'vendor.id': vendor.id
+  });
+    
+  if (existingChat) {
+    res.status(409).json({ message: 'Chat already exists', chat: existingChat });
+    return;
+  }
+
+  const chat = await Chat.create({
+    userObj,
+    vendorObj,
+    time: new Date(),
+    messages: [],
+    invoices: [],
+  });
+
+  const returnChat = await Chat.findById(chat._id)
+    .populate('user', 'name')
+    .populate('vendor', 'name');
+
+  res.status(201).json(returnChat);
+});
+
+ChatController.get('/:id/messages', verifyChatAccess, async (req, res) => {
+  const chat = req.chat;
+  if (!chat) {
+    res.status(500).json('Chat not found');
+    return; 
+  }
+  res.json(chat.messages.sort((a, b) => +new Date(a.time) - +new Date(b.time)));
+});
+
+ChatController.post('/:id/messages', verifyChatAccess, async (req, res) => {
+  const chat = req.chat;
+  const { text } = req.body;
+
+  if (!text) {
+    res.status(400).json('Missing message text');
+    return;
+  }
+
+  const newMessage: IMessage | any = {
+    text,
+    time: new Date(),
+   // sender: chat?.user.id === req.auth?.payload.sub ? 'user' : 'vendor',
+    user: {
+   //   id: chat?.user.id!,
+      name: chat?.user.name,
+      email: chat?.user.email,
+    },
+    vendor: {
+      id: chat?.vendor.id!,
+      name: chat?.vendor.name,
+      email: chat?.vendor.user.email,
+    }
+  };
```

Comment:
user and vendor refer to other models, need to actually those.

Please double check this suggestion before accepting.

Note that I am choosing to use chat.user._id here, so to accept this suggestion as written, the above comment I left needs to be resolved by filtering out undesired data including user._id at the endpoints, not in the middleware.
```suggestion

  // Should be set by verifyChatAccess middleware
  if (!chat) {
    res.status(400).send('Internal server error');
    return;
  }

  const newMessage: IMessage | any = {
    text,
    time: new Date(),
   // sender: chat?.user.id === req.auth?.payload.sub ? 'user' : 'vendor',
    user: chat.user._id,
    vendor: chat.vendor._id,
  };
```

Reply:
Just noticed that chat is already checked, just after the new message is created.